src/session.ts

```diff
@@ -520,20 +527,87 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-function hold(this: WebPhoneSession): Promise<any> {
+async function sendReinvite(this: WebPhoneSession, options: any = {}): Promise<any> {
+    if (options === void 0) {
+        options = {};
+    }
+    if (this.pendingReinvite) {
+        throw new Error('Reinvite in progress. Please wait until complete, then try again.');
+    }
+    if (!this.sessionDescriptionHandler) {
+        throw new Error("No SessionDescriptionHandler, can't send reinvite..");
+    }
+    this.pendingReinvite = true;
+    options.modifiers = options.modifiers || [];
+
+    return new Promise(async (resolve, reject) => {
+        try {
+            const description = await this.sessionDescriptionHandler.getDescription(
+                options.sessionDescriptionHandlerOptions,
+                options.modifiers
+            );
+            this.sendRequest(C.INVITE, {
+                body: description,
+                receiveResponse: (response: IncomingResponse) => {
+                    if (response.statusCode === 200) resolve(response);
+                    return this.receiveReinviteResponse(response);
+                }
+            });
+        } catch (e) {
+            this.pendingReinvite = false;
+            reject(e);
+        }
+    });
+}
+
+/*--------------------------------------------------------------------------------------------------------------------*/
+
+async function hold(this: WebPhoneSession): Promise<any> {
+    if (this.status !== Session.C.STATUS_WAITING_FOR_ACK && this.status !== Session.C.STATUS_CONFIRMED) {
+        throw new Exceptions.InvalidStateError(this.status);
+    }
+    if (this.localHold) {
+        throw new Error('Session already on hold');
+    }
     this.stopMediaStats();
-    return setLocalHold(this, true);
+    let options = {
+        modifiers: []
+    };
+    options.modifiers = options.modifiers || [];
+    options.modifiers.push(this.sessionDescriptionHandler.holdModifier);
+    try {
+        this.logger.log('Hold Initiated');
+        let response = await this._sendReinvite(options);
+        this.localHold = true;
+        this.logger.log('Hold completed: ' + response.body);
+    } catch (e) {
+        throw new Error('Hold could not be completed');
+    }
 }
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-function unhold(this: WebPhoneSession): Promise<any> {
-    return setLocalHold(this, false);
+async function unhold(this: WebPhoneSession): Promise<any> {
+    if (this.status !== Session.C.STATUS_WAITING_FOR_ACK && this.status !== Session.C.STATUS_CONFIRMED) {
+        throw new Exceptions.InvalidStateError(this.status);
+    }
+    if (!this.localHold) {
```

Comment:
Session can also be on hold when hold is initiated by another party. I don't think it's safe to rely on local var.

Reply:
This is going to be local for the session. I am pretty sure ,  the session put on hold by other party  will not affects localhold - i can reconfirm .